robot/packages/deferred_package_manager.go

```diff
@@ -0,0 +1,175 @@
+		establishConnection: establishConnection,
+		cloudManagerArgs:    cloudManagerConstructorArgs{cloudConfig, packagesDir, logger},
+		lastSyncedManager:   &noopManager,
+		logger:              logger,
+	}
+}
+
+// Sync syncs packages and removes any not in the list from the local file system.
+// If there are packages missing on the local fs, this will wait while attempting to establish a connection to app.viam.
+//
+// Sync is the core state-setting operation of the package manager so if we sync with one manager,
+// all subsequent operations should use the same manager until the next sync.
+func (m *deferredPackageManager) Sync(ctx context.Context, packages []config.PackageConfig) error {
+	m.lastSyncedManagerLock.Lock()
+	defer m.lastSyncedManagerLock.Unlock()
+	mgr, err := m.getManagerForSync(ctx, packages)
+	if err != nil {
+		return err
+	}
+	m.lastSyncedManager = mgr
+	return mgr.Sync(ctx, packages)
+}
+
+// Cleanup removes all unknown packages from the working directory.
+func (m *deferredPackageManager) Cleanup(ctx context.Context) error {
+	m.lastSyncedManagerLock.Lock()
+	defer m.lastSyncedManagerLock.Unlock()
+	return m.lastSyncedManager.Cleanup(ctx)
+}
+
+// PackagePath returns the package if it exists and already download. If it does not exist it returns a ErrPackageMissing error.
+func (m *deferredPackageManager) PackagePath(name PackageName) (string, error) {
+	m.lastSyncedManagerLock.Lock()
+	defer m.lastSyncedManagerLock.Unlock()
+	return m.lastSyncedManager.PackagePath(name)
+}
+
+// Close manager.
+func (m *deferredPackageManager) Close(ctx context.Context) error {
+	m.lastSyncedManagerLock.Lock()
+	defer m.lastSyncedManagerLock.Unlock()
+	return m.lastSyncedManager.Close(ctx)
+}
+
+// getManagerForSync returns the cloudManager if there is one cached (or if there are missing packages)
+// otherwise return noopManager and async get a cloudManager.
+func (m *deferredPackageManager) getManagerForSync(ctx context.Context, packages []config.PackageConfig) (ManagerSyncer, error) {
+	m.cloudManagerLock.Lock()
+	// the lock is handed to the goroutine so we do not defer an unlock
+
+	// return cached cloud manager if possible
+	if m.cloudManager != nil {
+		m.cloudManagerLock.Unlock()
+		return m.cloudManager, nil
+	}
+
+	// if we are missing packages, run createCloudManager synchronously
+	if m.isMissingPackages(packages) {
+		mgr, err := m.createCloudManager(ctx)
+		if err == nil {
+			// err == nil, not != nil
+			m.cloudManager = mgr
+			m.logger.Info("cloud package manager created synchronously")
+		}
+		m.cloudManagerLock.Unlock()
+		return mgr, err
+	}
+
+	// otherwise, spawn a goroutine to establish the connection and use a noopManager in the meantime
+	// hold the cloudManagerLock until this finishes
+	goutils.PanicCapturingGo(func() {
+		defer m.cloudManagerLock.Unlock()
+		mgr, err := m.createCloudManager(ctx)
+		if err != nil {
+			m.logger.Warnf("failed to create cloud package manager %v", err)
```

Comment:
This is pretty noisy I think. In this case we have all packages we need and the robot is offline. How often does it happen, every 5sec? I'm concerned if we have this a ton of logs will be uploaded when the robot reconnects.

Reply:
If there is an error, and the robot is using a cloud config, yes, this will print once every reconfigure.

In my personal testing, it didn't seem terribly noisy (compared to other sources of noise).

What about downgrading this to an info? Given the newness of this, I'd vote that we keep this message in here until at least a few stable versions have passed and we are confident in this code